File: outco/curriculum/target_practice_solutions/py/w3_d3_matrix.py
#
# Target Practice - Matrices
#
#
# Problem 1:  Robot Paths
#
# Prompt:   Given a matrix of zeroes, determine how many unique paths exist
#           from the top left corner to the bottom right corner
#
# Input:    An Array of Array of Integers (matrix)
# Output:   Integer
#
# Example:  matrix = [[0,0,0,0],
#                     [0,0,0,0],
#                     [0,0,0,0]]
#
#           robotPaths(matrix) = 38
#
#
#           matrix = [[0,0,0],
#                     [0,0,0]]
#
#           robotPaths(matrix) = 4

# Note:     From any point, you can travel in the four cardinal directions
#           (north, south, east, west). A path is valid as long as it travels
#           from the top left corner to the bottom right corner, does not go
#           off of the matrix, and does not travel back on itself


def robot_paths(matrix):
    pass
    def traverse(x, y):
        if (x < 0 or y < 0 or x >= len(matrix[0]) or y >= len(matrix)):
            return 0
        elif matrix[y][x] == 1:
            return 0
        elif (x == len(matrix[0]) - 1 and y == len(matrix) - 1):
            return 1
        matrix[y][x] = 1
        sum = 0
        sum += traverse(x + 1, y)
        sum += traverse(x - 1, y)
        sum += traverse(x, y + 1)
        sum += traverse(x, y - 1)
        matrix[y][x] = 0
        return sum

    return traverse(0, 0)


# Problem # 2: Matrix spiral
#
# Given an (MxN) matrix of integers, return an array in spiral order.
#
# Input:  matrix {Integer[][]}
# Output: {Integer}
#
#
# Example:
# Input:  [[ 1, 2, 3],
#          [ 4, 5, 6],
#          [ 7, 8, 9]]
# Output: [1, 2, 3, 6, 9, 8, 7, 4, 5]
#
#

# Time Complexity: O(MN)
# Auxiliary Space Complexity: O(MN)

def matrix_spiral(input):
    pass
    output = []

    row_min, col_min = 0, 0
    row_max, col_max = len(input) - 1, len(input[0]) - 1

    while row_min <= row_max and col_min <= col_max:
        for col in range(col_min, col_max + 1):
            output.append(input[row_min][col])
        row_min += 1

        for row in range(row_min, row_max + 1):
            output.append(input[row][col_max])
        col_max -= 1

        if row_min <= row_max:
            for col in range(col_max, col_min-1, -1):
                output.append(input[row_max][col])
            row_max -= 1

        if col_min <= col_max:
            for row in range(row_max, row_min - 1, -1):
                output.append(input[row][col_min])
            col_min += 1

    return output

#############################################
########  DO NOT TOUCH TEST BELOW!!!  #######
#############################################

def expect(count, name, test):
    if (count == None or not isinstance(count, list) or len(count) != 2):
        count = [0, 0]
    else:
        count[1] += 1

    result = 'false'
    errMsg = None
    try:
        if test():
            result = ' true'
            count[0] += 1
    except Exception as err:
        errMsg = str(err)

    print('  ' + (str(count[1]) + ')   ') + result + ' : ' + name)
    if errMsg != None:
        print('       ' + errMsg + '\n')

def lists_equal(lst1, lst2):
    if len(lst1) != len(lst2):
        return False
    for i in range(0, len(lst1)):
        if lst1[i] != lst2[i]:
            return False
    return True

print('Robot Paths Tests')
test_count = [0, 0]


def test():
    matrix = [[0, 0, 0, 0],
              [0, 0, 0, 0],
              [0, 0, 0, 0]]
    example = robot_paths(matrix)
    return example == 38


expect(test_count, 'should work on first example input', test)


def test():
    matrix = [[0, 0, 0],
              [0, 0, 0]]
    example = robot_paths(matrix)
    return example == 4


expect(test_count, 'should work on second example input', test)


def test():
    matrix = [[0]]
    example = robot_paths(matrix)
    return example == 1


expect(test_count, 'should work on single-element input', test)


def test():
    matrix = [[0, 0, 0, 0, 0, 0]]
    example = robot_paths(matrix)
    return example == 1


expect(test_count, 'should work on single-row input', test)


def test():
    matrix = [[0],
              [0],
              [0],
              [0],
              [0]]
    example = robot_paths(matrix)
    return example == 1


expect(test_count, 'should work on a 5 x 8 matrix input', test)


def test():
    matrix = [[0, 0, 0, 0, 0, 0, 0, 0],
              [0, 0, 0, 0, 0, 0, 0, 0],
              [0, 0, 0, 0, 0, 0, 0, 0],
              [0, 0, 0, 0, 0, 0, 0, 0],
              [0, 0, 0, 0, 0, 0, 0, 0]]
    print("  Please be patient, test 6 may take longer to run")
    example = robot_paths(matrix)
    return example == 7110272


print('PASSED: ' + str(test_count[0]) + ' / ' + str(test_count[1]) + '\n\n')

print("\nMatrix Spiral")
test_count = [0, 0]


def test():
    matrix = [[]]
    example = matrix_spiral(matrix)
    return example == []


expect(test_count, 'should work on empty matrix input', test)


def test():
    matrix = [[1]]
    example = matrix_spiral(matrix)
    return example == [1]


expect(test_count, "should work on single element matrix input", test)


def test():
    matrix = [[1],
              [2],
              [3],
              [4],
              [5]]
    example = matrix_spiral(matrix)
    return example == [1, 2, 3, 4, 5]


expect(test_count, "should work on single column matrix input", test)


def test():
    matrix = [[1, 2],
              [4, 3]]
    example = matrix_spiral(matrix)
    return example == [1, 2, 3, 4]


expect(test_count, "should work on square matrix input", test)


def test():
    matrix = [[1, 2, 3, 4]]
    example = matrix_spiral(matrix)
    return example == [1, 2, 3, 4]


expect(test_count, "should work on single row matrix input", test)


def test():
    matrix = [[ 1,  2,  3],
              [ 4,  5,  6],
              [ 7,  8,  9],
              [10, 11, 12],
              [13, 14, 15]]
    example = matrix_spiral(matrix)
    return example == [1, 2, 3, 6, 9, 12, 15, 14, 13, 10, 7, 4, 5, 8, 11]


expect(test_count, "should work on 3 x 5 matrix input", test)


def test():
    matrix = [[  1,  2,  3,  4, 5],
              [ 12, 13, 14, 15, 6],
              [ 11, 10,  9,  8, 7]]
    example = matrix_spiral(matrix)
    return example == [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]


expect(test_count, "should work on 5 x 3 matrix input", test)

print('PASSED: ' + str(test_count[0]) + ' / ' + str(test_count[1]) + '\n\n')

File: outco/curriculum/target_practice_solutions/cs/w3_d3_matrix.cs
/*
 * Target Practice - Matrices
 *
 *  Problem 1:  Robot Paths
 *
 *  Prompt:   Given a matrix of zeroes, determine how many unique paths exist
 *            from the top left corner to the bottom right corner
 *
 *  Input:    An Array of Array of ints (matrix)
 *  Output:   int
 *
 *  Example:  matrix = [[0,0,0,0],
 *                      [0,0,0,0],
 *                      [0,0,0,0]]
 *
 *            robotPaths(matrix) = 38
 *
 *
 *            matrix = [[0,0,0],
 *                      [0,0,0]]
 *
 *            robotPaths(matrix) = 4
 *
 *  Note:     From any point, you can travel in the four cardinal directions
 *            (north, south, east, west). A path is valid as long as it travels
 *            from the top left corner to the bottom right corner, does not go
 *            off of the matrix, and does not travel back on itself
 */

using System;
using System.Linq;
using System.Collections.Generic;

class Problems {

  public static int robotPaths(int[,] matrix) {
    return traverse(0, 0, matrix);
  }

  private static int traverse(int row, int col, int[,] matrix) {
    if(row < 0 || col < 0 || row >= matrix.GetLength(0) || col >= matrix.GetLength(1)) {
      return 0;
    } else if (matrix[row,col] == 1) {
      return 0;
    } else if (col == matrix.GetLength(1) - 1 && row == matrix.GetLength(0) - 1) {
      return 1;
    }
    matrix[row,col] = 1;
    int sum = 0;
    sum += traverse(row+1, col, matrix);
    sum += traverse(row-1, col, matrix);
    sum += traverse(row, col+1, matrix);
    sum += traverse(row, col-1, matrix);
    matrix[row,col] = 0;
    return sum;
   }


 /*
  *  Problem 2: Matrix Spiral
  *
  *  Given an (MxN) matrix of integers, return an array in spiral order.
  *
  *
  *  Input:  matrix {int[,]}
  *  Output: {int}
  *
  *
  * Example:
  *  Input:  [[ 1, 2, 3],
  *           [ 4, 5, 6],
  *           [ 7, 8, 9]]
  *
  *  Output: [1, 2, 3, 6, 9, 8, 7, 4, 5]
  *
  */


 // Time Complexity: O(MN)
 // Auxiliary Space Complexity: O(MN)

 public static int[] matrixSpiral(int[,] matrix) {
   if(matrix.GetLength(0) == 0) {
     return new int[0];
   }

   int[] results = new int[matrix.GetLength(0) * matrix.GetLength(1)];
   int rowMin = 0;
   int colMin = 0;
   int rowMax = matrix.GetLength(0) - 1;
   int colMax = matrix.GetLength(1) - 1;
   int counter = 0;
   int i;

   while (colMin <= colMax && rowMin <= rowMax) {
       for(i = colMin; i <= colMax; i++){
           results[counter] = matrix[rowMin,i];
           counter++;
       }
       rowMin++;
       for(i = rowMin; i <= rowMax; i++){
           results[counter] = matrix[i,colMax];
           counter++;
       }
       colMax--;
       if(rowMin <= rowMax){
           for(i = colMax; i >= colMin; i--){
               results[counter] = matrix[rowMax,i];
               counter++;
           }
           rowMax--;
       }
       if(colMin <= colMax){
           for(i = rowMax; i >= rowMin; i--){
               results[counter] = matrix[i,colMin];
               counter++;
           }
           colMin++;
       }
     }

   return results;
 }

}

 ////////////////////////////////////////////////////////////
 ///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
 ////////////////////////////////////////////////////////////

// use the Test class to run the test cases
class Test{

  public static void Main() {
    robotPathsTests();
    matrixSpiralTests();
	}

  private static void robotPathsTests() {
    int[] testCount = {0, 0};
    Console.WriteLine("Robot Paths Tests");
    runTest(robotPathsTest1, "should work on example input", testCount);
    runTest(robotPathsTest2, "should work on second example input", testCount);
    runTest(robotPathsTest3, "should work on single-element input", testCount);
    runTest(robotPathsTest4, "should work on single-row input", testCount);
    runTest(robotPathsTest5, "should work on single-column input", testCount);
    runTest(robotPathsTest6, "should work on a 5 x 8 matrix input", testCount);
    printTestsPassed(testCount);
  }

  private static void matrixSpiralTests() {
    int[] testCount = {0, 0};
    Console.WriteLine("Matrix Spiral Tests");
    runTest(matrixSpiralTest1, "should work for empty matrix input", testCount);
    runTest(matrixSpiralTest2, "should work for single element matrix input", testCount);
    runTest(matrixSpiralTest3, "should work for single column matrix input", testCount);
    runTest(matrixSpiralTest4, "should work for square matrix input", testCount);
    runTest(matrixSpiralTest5, "should work for single row matrix input", testCount);
    runTest(matrixSpiralTest6, "should work on 3 x 5 matrix input", testCount);
    runTest(matrixSpiralTest7, "should work on 5 x 3 matrix input", testCount);
    printTestsPassed(testCount);
  }


  private static bool robotPathsTest1() {
    int[,] matrix = {{0,0,0,0},
                          {0,0,0,0},
                          {0,0,0,0}};
    return Problems.robotPaths(matrix) == 38;
  }

  private static bool robotPathsTest2() {
    int[,] matrix = {{0,0,0},
                          {0,0,0}};
    return Problems.robotPaths(matrix) == 4;
  }

  private static bool robotPathsTest3() {
    int[,] matrix = {{0}};
    return Problems.robotPaths(matrix) == 1;
  }

  private static bool robotPathsTest4() {
    int[,] matrix = {{0,0,0,0,0,0}};
    return Problems.robotPaths(matrix) == 1;
  }

  private static bool robotPathsTest5() {
    int[,] matrix = {{0},
                          {0},
                          {0},
                          {0},
                          {0}};
    return Problems.robotPaths(matrix) == 1;
  }

  private static bool robotPathsTest6() {
    int[,] matrix = {{0,0,0,0,0,0,0,0},
                          {0,0,0,0,0,0,0,0},
                          {0,0,0,0,0,0,0,0},
                          {0,0,0,0,0,0,0,0},
                          {0,0,0,0,0,0,0,0}};
    Console.WriteLine("  Please be patient, test 6 may take longer to run");
    return Problems.robotPaths(matrix) == 7110272;
  }


  private static bool matrixSpiralTest1() {
    int[,] matrix = {{}};
    int[] test = Problems.matrixSpiral(matrix);
    int[] expected = {};
    return test.SequenceEqual(expected);
  }

  private static bool matrixSpiralTest2() {
    int[,] matrix = {{1}};
    int[] test = Problems.matrixSpiral(matrix);
    int[] expected = {1};
    return test.SequenceEqual(expected);
  }

  private static bool matrixSpiralTest3() {
    int[,] matrix = {{1},
                          {2},
                          {3},
                          {4},
                          {5}};
    int[] test = Problems.matrixSpiral(matrix);
    int[] expected = {1, 2, 3, 4, 5};
    return test.SequenceEqual(expected);
  }

  private static bool matrixSpiralTest4() {
    int[,] matrix = {{1, 2},
                          {4, 3}};
    int[] test = Problems.matrixSpiral(matrix);
    int[] expected = {1, 2, 3, 4};
    return test.SequenceEqual(expected);
  }

  private static bool matrixSpiralTest5() {
    int[,] matrix = {{1, 2, 3, 4}};
    int[] test = Problems.matrixSpiral(matrix);
    int[] expected = {1, 2, 3, 4};
    return test.SequenceEqual(expected);
  }

  private static bool matrixSpiralTest6() {
    int[,] matrix = {{1, 2, 3},
                      {4, 5, 6},
                      {7, 8, 9},
                      {10, 11, 12},
                      {13, 14, 15}};
    int[] test = Problems.matrixSpiral(matrix);
    int[] expected = {1, 2, 3, 6, 9, 12, 15, 14, 13, 10, 7, 4, 5, 8, 11};
    return test.SequenceEqual(expected);
  }

  private static bool matrixSpiralTest7() {
    int[,] matrix = {{1, 2, 3, 4, 5},
                      {12, 13, 14, 15, 6},
                      {11, 10, 9, 8, 7}};
    int[] test = Problems.matrixSpiral(matrix);
    int[] expected = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};
    return test.SequenceEqual(expected);
  }


  // DO NOT TOUCH FUNCTIONS BELOW

  // Custom runTest function to handle tests
  // Function<bool> test : performs a set of operations and returns a bool
  //   indicating if test passed
  // string name : describes the test
  // int[] testCount : keeps track out how many tests pass and how many total
  //   in the form of a two item array i.e., [0, 0]
  private static void runTest(Func<bool> test, string testName, int[] testCount){
      testCount[1]++;
      bool testPassed = false;
      // Attempt to run test and suppress exceptions on failure
      try {
          testPassed = test();
          if(testPassed) testCount[0]++;
      } catch (Exception e) {
        Console.WriteLine(e);
      }
      string result = "  " + (testCount[1] + ")   ") + testPassed + " : " + testName;
      Console.WriteLine(result);
  }
  // this is to print the number of test cases passed for a test suite function
  private static void printTestsPassed(int[] testCount) {
    Console.WriteLine("PASSED: " + testCount[0] + " / " + testCount[1] + "\n\n");
  }
}
